File: c#/BigEgg.Algorithm/BigEgg.Algorithm/Collections/ResizingArrayBag.cs
﻿using System;
using System.Collections;
using System.Collections.Generic;

namespace BigEgg.Algorithm.Collections
{
    public class ResizingArrayBag<Item> : IBag<Item>
    {
        private Item[] a;
        private int N;

        public ResizingArrayBag()
        {
            N = 0;
            a = new Item[2];
        }

        /// <summary>
        /// Is this bag empty?
        /// </summary>
        /// <returns><c>True</c> if this bag is empty; <c>false</c> otherwise</returns>
        public bool IsEmpty()
        {
            return N == 0;
        }

        /// <summary>
        /// Returns the number of items in the bag.
        /// </summary>
        /// <returns>The number of items in the bag.</returns>
        public int Size()
        {
            return N;
        }

        private void Resize(int capacity)
        {
            Item[] temp = new Item[capacity];
            for (int i = 0; i < N; i++)
            {
                temp[i] = a[i];
            }
            a = temp;
        }

        /// <summary>
        /// Adds the item to this bag.
        /// </summary>
        /// <param name="item">The item to add.</param>
        /// <exception cref="System.ArgumentNullException">Item cannot be null.</exception>
        public void Add(Item item)
        {
            if (item == null) { throw new ArgumentNullException(); }
            if (N == a.Length) Resize(2 * a.Length);
            a[N++] = item;
        }

        /// <summary>
        /// Returns a string representation of this bag.
        /// </summary>
        /// <returns>The sequence of items in the bag, separated by spaces.</returns>
        public override string ToString()
        {
            return String.Join(" ", this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through the bag.
        /// </summary>
        /// <returns>
        /// A <see cref="T:System.Collections.Generic.IEnumerator`1" /> that can be used to iterate through the bag.
        /// </returns>
        public IEnumerator<Item> GetEnumerator()
        {
            return new Enumerator(this);
        }

        /// <summary>
        /// Returns an enumerator that iterates through a bag.
        /// </summary>
        /// <returns>
        /// An <see cref="T:System.Collections.IEnumerator" /> object that can be used to iterate through the bag.
        /// </returns>
        IEnumerator IEnumerable.GetEnumerator()
        {
            return new Enumerator(this);
        }

        private class Enumerator : IEnumerator<Item>, IDisposable, IEnumerator
        {
            private int i = -1;
            private int N;
            private Item[] a;

            public Enumerator(ResizingArrayBag<Item> bag)
            {
                a = bag.a;
                N = bag.N;
            }

            public Item Current
            {
                get { return a[i]; }
            }

            public void Dispose()
            {
                a = null;
                N = 0;
            }

            object IEnumerator.Current
            {
                get { return a[i]; }
            }

            public bool MoveNext()
            {
                return ++i < N;
            }

            public void Reset()
            {
                i = -1;
            }
        }
    }
}

File: java/Algorithms/src/Collections/ResizingArrayBag.java
package Collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

/**
 * Created by jianming.xiao on 10/12/14.
 */
public class ResizingArrayBag<Item> implements IBag<Item> {
    private Item[] a;
    private int N;

    public ResizingArrayBag() {
        a = (Item[]) new Object[2];
        N = 0;
    }

    /**
     * Is this bag empty?
     *
     * @return true if this bag is empty; false otherwise
     */
    public boolean isEmpty() {
        return N == 0;
    }

    /**
     * Returns the number of items in this bag.
     *
     * @return the number of items in this bag
     */
    public int size() {
        return N;
    }

    /**
     * Resize the underlying array holding the elements
     *
     * @param capacity the new capacity of the bag
     */
    private void resize(int capacity) {
        Item[] temp = (Item[]) new Object[capacity];
        for (int i = 0; i < N; i++) {
            temp[i] = a[i];
        }
        a = temp;
    }

    /**
     * Adds the item to this bag.
     *
     * @param item the item to add to this bag
     * @throws NullPointerException if the <tt>item</tt> is null;
     */
    public void add(Item item) {
        if (item == null) throw new NullPointerException();
        if (N == a.length) resize(a.length * 2);
        a[N++] = item;
    }

    /**
     * Returns a string representation of this bag.
     *
     * @return the sequence of items, separated by spaces
     */
    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();
        for (Item item : this) {
            stringBuilder.append(item.toString()).append(" ");
        }
        return stringBuilder.toString();
    }

    /**
     * Returns an iterator that iterates over the items in this bag.
     *
     * @return an iterator that iterates over the items in this bag
     */
    public Iterator<Item> iterator() {
        return new ResizingArrayBagIterator();
    }

    private class ResizingArrayBagIterator implements Iterator<Item> {
        private int currnet = 0;

        public boolean hasNext() {
            return currnet < N;
        }

        public Item next() {
            if (!hasNext()) throw new NoSuchElementException();
            return a[currnet++];
        }

        public void remove() {
            throw new UnsupportedOperationException();
        }
    }
}
